web/src/client/http.js

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright (c) [2024] SUSE LLC
+ *
+ * All Rights Reserved.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of version 2 of the GNU General Public License as published
+ * by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, contact SUSE LLC.
+ *
+ * To contact SUSE LLC about this file by physical or electronic mail, you may
+ * find current contact information at www.suse.com.
+ */
+
+// @ts-check
+
+/**
+ * @callback RemoveFn
+ * @return {void}
+ */
+
+/**
+ * Agama WebSocket client.
+ *
+ * Connects to the Agama WebSocket server and reacts on the events.
+ * This class is not expected to be used directly, but through the
+ * HTTPClient API.
+ */
+class WSClient {
+  /**
+   * @param {URL} url - Websocket URL.
+   */
+  constructor(url) {
+    this.client = new WebSocket(url.toString());
+    this.client.onmessage = (event) => {
+      this.dispatchEvent(event);
+    };
+    this.handlers = [];
+  }
+
+  /**
+   * Registers a handler for events.
+   *
+   * The handler is executer for all the events. It is up to the callback to
+   * filter the relevant events.
+   *
+   * @param {(object) => void} func - Handler function to register.
+   * @return {RemoveFn}
+   */
+  onEvent(func) {
+    this.handlers.push(func);
+    return () => {
+      const position = this.handlers.indexOf(func);
+      if (position > -1) this.handlers.splice(position, 1);
+    };
+  }
+
+  /**
+   * @private
+   *
+   * Dispatchs an event by running all the handlers.
+   *
+   * @param {object} event - Event object, which is basically a websocket message.
+   */
+  dispatchEvent(event) {
+    const eventObject = JSON.parse(event.data);
+    this.handlers.forEach((f) => f(eventObject));
+  }
+}
+
+/**
+ * Agama HTTP API client.
+ */
+class HTTPClient {
+  /**
+   * @param {URL} url - URL of the HTTP API.
+   */
+  constructor(url) {
+    const httpUrl = new URL(url.toString());
```

Comment:
Because `HTTPClient` and `WSClient` receive an `URL` object (instead of a string). In this case, I am converting such a `URL` into a string (because `fetch` uses a string).

Reply:
I missed something. You're manipulating the url in lines 91 and 92 without converting it to string. However, in line 86 you have converted it to string, manipuladed in line 87 and then converted to string again in line 88.

Reply:
oh, I see now, the first `toString` call is not needed.

Reply:
Actually, I decided to use the `toString` method in both places because, [according to the docs](https://developer.mozilla.org/en-US/docs/Web/API/URL), the constructor is expected to get a string.